pkg/provision/workspace/serviceaccountToken.go

```diff
@@ -0,0 +1,146 @@
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/utils/pointer"
+
+	"github.com/devfile/devworkspace-operator/pkg/common"
+)
+
+func ProvisionServiceAccountTokensInto(podAdditions *v1alpha1.PodAdditions, workspace *common.DevWorkspaceWithConfig) error {
+	saTokenVolumeMounts, saTokenVolumes, err := getSATokensVolumesAndVolumeMounts(workspace.Config.Workspace.ServiceAccount.ServiceAccountTokens)
+	if err != nil {
+		return err
+	}
+	podAdditions.VolumeMounts = append(podAdditions.VolumeMounts, saTokenVolumeMounts...)
+	podAdditions.Volumes = append(podAdditions.Volumes, saTokenVolumes...)
+	return nil
+}
+
+// Returns VolumeMounts and (projected) Volumes corresponding to the provided ServiceAccount tokens.
+// An error is returned if at least two ServiceAccount tokens share the same mounth path and path.
+func getSATokensVolumesAndVolumeMounts(serviceAccountTokens []v1alpha1.ServiceAccountToken) ([]corev1.VolumeMount, []corev1.Volume, error) {
+	volumeMounts := []corev1.VolumeMount{}
+	volumes := []corev1.Volume{}
+	// Map of mountPath -> ServiceAccountTokens, to detect colliding volumeMounts
+	mountPathToSATokens := map[string][]v1alpha1.ServiceAccountToken{}
+
+	// Ordered list of mountPaths to process, to avoid random iteration order on maps
+	var mountPathOrder []string
+	for _, saToken := range serviceAccountTokens {
+		if len(mountPathToSATokens[saToken.MountPath]) == 0 {
+			mountPathOrder = append(mountPathOrder, saToken.MountPath)
+		}
+		mountPathToSATokens[saToken.MountPath] = append(mountPathToSATokens[saToken.MountPath], saToken)
+	}
+	sort.Strings(mountPathOrder)
+
+	for _, mountPath := range mountPathOrder {
+		saTokens := mountPathToSATokens[mountPath]
+		vm, vol, err := generateSATokenProjectedVolume(mountPath, saTokens)
+		if err != nil {
+			return nil, nil, err
+		}
+		volumeMounts = append(volumeMounts, *vm)
+		volumes = append(volumes, *vol)
+	}
+	return volumeMounts, volumes, nil
+}
+
+// Returns a VolumeMount and projected Volume for the provided ServiceAccount tokens that share the same mount path.
+// The VolumeMount's mount path is set to the provided mount path and its name is set to the ServiceAccount token's name
+// if only a single ServiceAccount token is provided, otherwise a common VolumeMount name is used
+// if multiple ServiceAccount token's are provided.
+// An error is returned if at least two ServiceAccount tokens share the same mounth path and path.
+func generateSATokenProjectedVolume(mountPath string, saTokens []v1alpha1.ServiceAccountToken) (*corev1.VolumeMount, *corev1.Volume, error) {
+	// Check if two tokens share the same path and mountPath, which is invalid
+	err := checkSATokenPathCollisions(saTokens, mountPath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// If multiple tokens are being projected into the same mount path, use a common name
+	volumeName := common.ServiceAccountTokenProjectionName(mountPath)
+	if len(saTokens) == 1 {
+		volumeName = saTokens[0].Name
+	}
```

Comment:
nitpick for clarity:
```suggestion
	var volumeName string
	if len(saTokens) == 1 {
		volumeName = saTokens[0].Name
	} else {
		volumeName = common.ServiceAccountTokenProjectionName(mountPath)
	}
```
something about calling a function and then optionally discarding its result looks weird to me.

Or alternatively, maybe it makes sense to use the common name for all SA tokens and not try to give them a "nice" name?